### PR TITLE
Use the WAF as dynamic library

### PR DESCRIPTION
**Pull request by robertpi**
Remove the code to merge the .lib / C++ coee into the
ClrProfiler.Native.[dll|so|dynlib].

This means we now need to be able to find the plaform specific library,
so I've added some code for this. We use the environment variables to
find where the native library might be installed and load it from this
location.

Ensure that if the native library doesn't load for any reason the client
app can contiune functioning without security.

Fix the sample apps and security integrarion tests so they model a
datadog install in reasonably realistic way and that the integration
tests continue to work.

Change the way the rules file is stored. A resource file is probably
better than a very large constant.

Also: add back missing code from asp.net mvc scope creation

Fixes #

Changes proposed in this pull request:

@DataDog/apm-dotnet

### CODE REVIEW

**Review by andrewlock — COMMENTED** [review]
Just some minor comments, I'll defer to others for more practical reviews!

File: test/test-applications/Directory.Build.props
Comment (robertpi) [reply test/test-applications/Directory.Build.props:34]: Changed it back, I think it was something of an experiment that wasn't needed in the end.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:22]: The conditional logic repeated in the construct is because tuple deconstruction is not support in .NET framework 4.5. I have come up with a way to remove the conditional compilation in the construct, but I can't see anyway of avoiding it here.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:41]: It's a bit annoying that expression style switch doesn't allow regroup multiple cases together, so I've move to the old statement based switch which does allow cases to be regrouped.

It also reminded me that throwing isn't really a good idea in a static constructor, as it will mean any member accessed after static constructor failure will also throw, so I've changed that as well.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:93]: Yes, dotnet core on Windows supports post fixing the dll, that's documented here: https://docs.microsoft.com/en-us/dotnet/standard/native-interop/cross-platform

What surprises me that .NET Framework consuming a .NET standard library seems able to do this, or an least our integration tests seem to test this scenario and it seems to be working.

File: src/Datadog.Trace/AppSec/Waf/rule-set.txt
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/rule-set.txt:6]: Yes, no problem.

It's reasonably unlikely we'll ever need to change this file, it's provided by the app sec security team (who also created the native library we're loading)

It's also unlikely that it'll stay as part of the repository long term, the plan is that once the fleet management agent remote control protocol is implemented these rules will be provided by the datadog back end, via this new service.

**Review by anna-git — APPROVED** [review]

File: Datadog.Trace.Security.sln
Comment (lucaspimentel) [inline Datadog.Trace.Security.sln:40]: This looks like a typo. `EndProjectSectionEndProject` => `EndProjectSection` and `EndProject`

File: src/Datadog.Trace/AppSec/Security.cs
Comment (lucaspimentel) [inline src/Datadog.Trace/AppSec/Security.cs:40]: We have a (ridiculous long but optimized) `String.ToBoolean()` extension method which allows for different truthy values (`true`, `T`, `yes`, `Y`, `1`). We try to use it for all boolean configuration.

https://github.com/DataDog/dd-trace-dotnet/blob/b68cdfef714b378cd61ced1fa2894e0e215d90a9/src/Datadog.Trace/ExtensionMethods/StringExtensions.cs#L37

Comment (lucaspimentel) [reply src/Datadog.Trace/AppSec/Security.cs:40]: Is `DD_ENABLE_SECURITY` the final name for this across languages? In APM most "enabled" settings follow the `DD_{FEATURE}_ENABLED` pattern.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (lucaspimentel) [inline src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:99]: It's `Datadog`, not `DataDog` 😉 

(Yes, it's https://github.com/DataDog on GitHub, but nobody dares change it because it could break case sensitive package managers.)

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (lucaspimentel) [inline src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:118]: `COR_PROFILER_PATH_64` and `COR_PROFILER_PATH_32` are also supported on newer versions of .NET Framework. I could never find documentation of the exact version, but it was added around the 4.6 time frame.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (lucaspimentel) [inline src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:228]: We only support `Win32NT` out of these four.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (lucaspimentel) [inline src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:215]: [The documentation](https://docs.microsoft.com/en-us/dotnet/api/system.platformid) says this value was only returned in Silverlight. .NET Core on macOS returns `Unix`. You want want to use you `FrameworkDescription` helper instead.
https://github.com/DataDog/dd-trace-dotnet/blob/beba073dc8f9dd86b5f9becdb19feb4bdb8b69a0/src/Datadog.Trace/FrameworkDescription.cs#L16

File: src/Datadog.Trace/AppSec/Security.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Security.cs:40]: It is not the final version, the final version is DD_APPSEC_ENABLED. I logged this jira to remind me to change it, https://datadoghq.atlassian.net/browse/APPSEC-747, buy I'll do it now, as you mention it.

Comment (robertpi) [reply src/Datadog.Trace/AppSec/Security.cs:40]: Also use the correct boolean converter now.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:99]: If only we'd written it in case insensitive VB.NET!

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:228]: Changed.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs:215]: fixed.

File: Datadog.Trace.Security.sln
Comment (robertpi) [reply Datadog.Trace.Security.sln]: Thanks, correct.

File: src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs
Comment (robertpi) [reply src/Datadog.Trace/AppSec/Waf/NativeBindings/Native.cs]: My idea had been simple to take any available environment variable, even if it wasn't were the process loaded the profiler from, we might find the library in the same location. However, this could create problems with loading the wrong version of the library, so I changed the method just to look at the env var that would be correct for the process type.

**Pull request merged**
